---
title: "Medulloblastoma: applying MultiPLIER"
output:   
  html_notebook: 
    toc: true
    toc_float: true
---

**J. Taroni 2018**

In this notebook we'll prep medulloblastoma expression data (luckily the 
metadata is already in pretty good shape!) and apply MultiPLIER/recount2 model.

We're using data from the following publication 
([`GSE37382`](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE37382) 
on GEO):

> Northcott PA, Shih DJ, Peacock J, et al. [Subgroup-specific structural variation across 1,000 medulloblastoma genomes.](https://dx.doi.org/10.1038/nature11327) _Nature._ 
2012;488(7409):49-56. 

It has been processed with `SCAN::SCANfast` through 
[refine.bio](https://www.refine.bio/).

## Set up

#### Libraries

```{r}
# we need this library to convert from Ensembl gene identifiers to gene symbols
# for use with PLIER
library(org.Hs.eg.db)
```

#### Functions

```{r}
# magrittr pipe
`%>%` <- dplyr::`%>%`
# need PrepExpressionDF
source(file.path("util", "test_LV_differences.R"))
# GetNewDataB
source(file.path("util", "plier_util.R"))
```

#### Directory setup

```{r}
# directory that holds all the gene expression matrices
exprs.dir <- file.path("data", "expression_data")
# directories specific to this notebook
plot.dir <- file.path("plots", "37")
dir.create(plot.dir, recursive = TRUE, showWarnings = FALSE)
results.dir <- file.path("results", "37")
dir.create(results.dir, recursive = TRUE, showWarnings = FALSE)
```

## Read in data

Expression data

```{r}
exprs.file <- file.path(exprs.dir, "GSE37382_SCAN.pcl")
exprs.df <- readr::read_tsv(exprs.file)
```

PLIER model

```{r}
recount.file <- file.path("data", "recount2_PLIER_data", 
                          "recount_PLIER_model.RDS")
recount.plier <- readRDS(recount.file)
```

## Gene identifier conversion and aggregation

We're going to use the default behavior of `mapIds` here, i.e., if there are
1:many mappings, just return the first.

```{r}
gene.symbols.df <- 
  AnnotationDbi::mapIds(org.Hs.eg.db, keys = exprs.df$Gene, 
                        column = "SYMBOL", keytype = "ENSEMBL") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("Ensembl")
colnames(gene.symbols.df)[2] <- "Symbol"
```

```{r}
head(gene.symbols.df)
```

Now we'll join this to the expression `data.frame`

```{r}
annot.exprs.df <- gene.symbols.df %>%
  dplyr::inner_join(exprs.df, by = c("Ensembl" = "Gene")) %>%
  dplyr::select(-Ensembl)
colnames(annot.exprs.df)[1] <- "Gene"
```

Are there any duplicate gene symbols?

```{r}
any(duplicated(annot.exprs.df$Gene))
```

Okay, aggregate with the custom function `PrepExpressionDF` and write to file.

```{r}
agg.exprs.df <- PrepExpressionDF(annot.exprs.df) %>%
  # remove any rows without gene ids
  dplyr::filter(!is.na(Gene))
agg.file <- file.path(exprs.dir, "GSE37382_mean_agg.pcl")
readr::write_tsv(agg.exprs.df, path = agg.file)
```

Remove the `data.frame` that are large and no longer necessary.

```{r}
rm(exprs.df, gene.symbols.df, annot.exprs.df)
```

## MultiPLIER

We need a matrix where the gene symbols are rownames.

```{r}
exprs.mat <- agg.exprs.df %>%
  tibble::column_to_rownames("Gene") %>%
  as.matrix()
```

Apply the model

```{r}
medulloblastoma.b <- GetNewDataB(exprs.mat = exprs.mat,
                                 plier.model = recount.plier)
```

Save to file

```{r}
b.file <- file.path(results.dir, "medulloblastoma_recount2_B.RDS")
saveRDS(medulloblastoma.b, b.file)
```

